### ISSUE TITLE
Add support for Location items (through Mapview widget)

### ISSUE THREAD

**Issue by kaikreuzer** [openhab/openhab.android#574]
The Android client is lacking one of the standard widgets of openHAB: The [Mapview widget](http://docs.openhab.org/configuration/sitemaps.html#element-type-mapview), which can be used to display a geo location for a "Location" item.

In Basic UI (and Paper UI), we use OSM maps as these are freely available and the best fit for open source usage that does not want to confront its users with any special terms&conditions.

I'd assume that for Android, it might be more natural to use Google maps, though, which should be fine because Android users anyhow already have accepted all kinds of T&Cs from Google.

I leave the decision on how to do it best to the experts here. With all the recent activity on the app, I just wanted to take the opportunity to report this missing feature in the hope that it can be added :-)

FTR: I'd also like to extend the "Mapview" support to refer to a group item. In such a case, all members that are locations should be rendered on a single map. But this is probably something for which we might also want to enhance the sitemap information in the REST API.

**Comment by mueller-ma** [comment]
There are some people running android without play services ;)

**Comment by kaikreuzer** [comment]
Maybe one could offer both options (and use Google maps as a default, if the play services are installed)?

Btw, a suitable location pin can be found [here](https://github.com/htreu/openhab-core/blob/8768ec4396d6a2aa043f380084949b5a0160c72d/bundles/org.openhab.ui.paperui/patch/web/img/logo_pointer.png) in case it is helpful.

**Comment by maniac103** [comment]
Using OSM for the FOSS variant and Google Maps for the Play Store version indeed sounds like a good idea.

**Comment by maniac103** [comment]
I implemented support for Google Maps on top of PR #650: https://github.com/maniac103/openhab.android/commit/cb519d5e78a10c01a224c898c168ad9c3100d89e

It looks like this:
![screenshot_1516882414](https://user-images.githubusercontent.com/781373/35387894-e1e3134c-01d1-11e8-8f1e-60e1895a2705.png)
![screenshot_1516882426](https://user-images.githubusercontent.com/781373/35387895-e1fed3b6-01d1-11e8-856d-a4edbea44036.png)

Thoughts about that UI?

**Comment by kaikreuzer** [comment]
Looks great! But why is this done as a modal window and not directly embedded in the page (as it is done for the web UIs, e.g. Basic UI)?

**Comment by maniac103** [comment]
> Looks great! But why is this done as a modal window and not directly embedded in the page (as it is done for the web UIs, e.g. Basic UI)?

Guess why I asked for UI feedback ;-) I didn't think of checking basic UI. One approach I can think of is embedding the MapView with a fixed height (say 150dp) into the page itself, but don't allow scrolling it, and opening the scrollable popup when clicking it. WDYT?

**Comment by maniac103** [comment]
Now it looks like this:

![screenshot_1516886223](https://user-images.githubusercontent.com/781373/35390081-7c068924-01da-11e8-8e45-9cec2907052e.png)

I assume the label (+ icon) can be killed off, right?

**Comment by kaikreuzer** [comment]
> I assume the label (+ icon) can be killed off, right?

Yes, I would even remove the location name ("Brandenburg Gate Berlin") as it rather wastes space.

Zoom level: Good question. If there is any default that Android/Maps chooses anyhow, go for that :-)

**Comment by maniac103** [comment]
> Yes, I would even remove the location name ("Brandenburg Gate Berlin") as it rather wastes space.

That's the label I was referring to :-)

**Comment by mueller-ma** [comment]
Looks good so far!

> Yes, I would even remove the location name ("Brandenburg Gate Berlin") as it rather wastes space.

I wouldn't. When you have multiple maps (e.g. for different devices) you dont know which marker is which device.
I know that you are planning to support multiple markers per map. What do you think how it should look in the smaller view? Different colored markers?

**Comment by kaikreuzer** [comment]
Would it be possible to put it on the marker as in [this screenshot]? This would then be the best solution for multiple markers as well.

**Comment by mueller-ma** [comment]
> This would then be the best solution for multiple markers as well.

Imagine 5 markers there, it would be a total chaos.

Maybe the map should take more space, 300dp? 500dp?

**Comment by kaikreuzer** [comment]
Well, maybe it could take more space depending on how many markers there are. For a single one, the above suggestion lgtm.

**Comment by maniac103** [comment]
> Would it be possible to put it on the marker as in this screenshot? This would then be the best solution for multiple markers as well.

That title appears when clicking a marker - which I currently have disabled for the in-list map as I felt that it's weird for the popup to only appear when clicking elsewhere on the map ... or am I wrong there?

**Comment by mueller-ma** [comment]
> or am I wrong there?

No, you're right. Otherwise people might not discover, that they can open a popup.

**Comment by maniac103** [comment]
@kaikreuzer Is the 'group Mapview' scenario possible with the current state of the rest API already? If so, can one be added to the demo sitemap for testing purposes? Thanks :-)

**Comment by kaikreuzer** [comment]
No, we don't yet have such a feature - from a user perspective I would suppose that one should simply want to provide a Group item as the item parameter for a Location widget. This is currently possible, but the UI/app would then have to do another call to the item resource to get hold of the group members. Ideally, the sitemap would directly expand the group and provide its members.

I have just created https://github.com/eclipse/smarthome/issues/4989 for it. Maybe something @lolodomo would be interested to look into?

**Comment by maniac103** [comment]
@kaikreuzer Ok, thanks for the confirmation. Indeed, it would be nice if the sitemap provided item for the group would provide the items of the group members as well ... there already is a `members` array in the `item` JSON object, was that one meant for that purpose originally?

As for the suggestions about the map (widget) label, I currently went for a middle ground:
![screenshot_1516955480](https://user-images.githubusercontent.com/781373/35431312-3b2d5bce-027c-11e8-98e5-e2a609a879fb.png)

Marker dragging to update the location is also supported now, which I used to move the location in the demo sitemap to its correct place again :-) I hope I took the `readOnly` information from the right position (the `readOnly` member of the item's `stateDescription` object, with default = false if it's missing)?

**Comment by kaikreuzer** [comment]
> there already is a members array in the item JSON object, was that one meant for that purpose originally?

I think this is there because it is a Group item - but by default the sitemaps do not recurse into group members and thus this list is simply empty.

> I currently went for a middle ground

Looks good, we can imho keep it that way!

>  I hope I took the readOnly information from the right position

Yes, sounds correct to me, thanks!

**Comment by mueller-ma** [comment]
> Also note that location items are potentially writable as well (if readOnly=false)

@kaikreuzer Is readOnly implemented in any oh snapshot? I couldnt find this option in the docs: https://docs.openhab.org/configuration/sitemaps.html#element-type-mapview

**Comment by maniac103** [comment]
I've updated #686 to add support for the newly-introduced location group item response:

![screenshot_1519907297](https://user-images.githubusercontent.com/781373/36844851-9f61df50-1d54-11e8-8c87-288e2f4f089a.png)

Unfortunately one can't color the markers (only alpha can be set), and only one info window can be shown at a given time, so AFAICT it's not possible to label all markers simultaneously.

**Comment by lolodomo** [comment]
Hi guys. 
Was the support of multiple markers made available for OSM or only Google Maps?

**Comment by mueller-ma** [comment]
Multiple markers are supported in both flavors.

**Comment by lolodomo** [comment]
With a unique URL for OSM? 
Would it be possible to do it the same way in Basic/Classic UI?

**Comment by mueller-ma** [comment]
We use a library for osm that allows setting multiple  markers.
AFAIK it's not possible to add multiple markers to a url, but there are some third party javascript libs.
